### DISCUSSION TITLE
Documentation update for DROPBOX_REMOTE_PATH environment variable

### DISCUSSION THREAD

**Discussion by shalafi99** [offen/docker-volume-backup#472] [General]
Hello,

I have set up dropbox for usage with docker-volume-backup today and noticed what appears to be an incorrect (or maybe just outdated because dropbox changed the API behavior, not sure) instruction inside the corresponding recipe/howto:

The dropbox "how-to" ([link](https://github.com/offen/docker-volume-backup/blob/main/docs/how-tos/set-up-dropbox.md)) states:

_Other parameters
Important: If you chose App folder access during the creation of your Dropbox app in step 1 above, you can only write in the app’s directory! This means, that DROPBOX_REMOTE_PATH must start with e.g. /Apps/YOUR_APP_NAME or /Apps/YOUR_APP_NAME/some_sub_dir_

As does the related "recipe" while referencing that same environment variable named DROPBOX_REMOTE_PATH

While setting up mine, I scoped permissions only to the app:
![image](https://github.com/user-attachments/assets/0f99a791-b5b5-466d-ba1a-148cd2e245ef)

Including the absolute path like instructed (/Apps/appname) resulted in this:

![image](https://github.com/user-attachments/assets/2983a1fa-4d59-4d9e-a18b-9e91695f653b)

I had to change it to DROPBOX_REMOTE_PATH=/diun (a relative path to the app) to achieve what would be the expected outcome:

level=INFO msg="Starting upload session for backup '/tmp/backup-diun-2024-09-16T16-18-37.tar.gz' at path '/diun'." storage=Dropbox
level=INFO msg="Uploaded a copy of backup '/tmp/backup-diun-2024-09-16T16-18-37.tar.gz' at path '/diun'." storage=Dropbox

![image](https://github.com/user-attachments/assets/ee5d4ea1-d3fb-4cf4-86c0-b2da7cdac759)
_(Note: I had not yet deleted the path from the previous run while taking the screenshot above)_

I have not done this before, but supposedly I could go for a PR to update "docs/recipes/index.md" and "docs/how-tos/set-up-dropbox.md" ?

Cheers

**Comment by m90** [comment]
I've never used the Dropbbox Storage (or even Dropbox) myself, so I can't tell if this is a change in API behavior or never worked as documented. Maybe @MaxJa4 who added this knows more about this?

**Comment by shalafi99** [comment]
I've noticed that in the "Configuration Reference" under github.io ([link](https://offen.github.io/docker-volume-backup/reference/)), the example value for DROPBOX_REMOTE_PATH appears to reflect current behavior:

_# DROPBOX_REMOTE_PATH="/my/directory"_

Also, there is no accompanying explanation text pointing out a need to include /Apps/appname in the variable (like it happens with the how-to / recipe).

**Reply by m90** [reply]
The original PR (#251) description says

> It is required to create an app in Dropbox. It is also important to set the DROPBOX_REMOTE_PATH to the app's subfolder, where the app has access to (scope)... unless when creating the app, one chooses to give the app full access to everything.

How is access scoped in your setup?

**Reply by shalafi99** [reply]
My app is using: "App folder– Access to a single folder created specifically for your app."

(_During app creation time, the other available choice was: "Full Dropbox– Access to all files and folders in a user's Dropbox."_)

It seems that not only the permissions are scoped to the app folder when but also file/namespace visibility.
I did some digging inside the Dropbox API documentation and found this [reference](https://www.dropbox.com/developers/reference/path-root-header-modes):

"Path Root Header Modes

When using the Dropbox User API, calls that access content are “rooted” to a particular directory per the apps permission. This means that relative paths are evaluated relative to this location, and paths outside of this location are not accessible. For apps that request App Folder access, these calls are rooted to the app folder. For applications that request Full Dropbox access, these calls are rooted to the User’s Dropbox folder by default.

Apps may also pass an additional HTTP header, Dropbox-API-Path-Root, in conjunction with authorization hearers to change where calls are rooted to, as allowed per the app’s permission. An application may wish to root to a frequently used shared folder for convenience, but the most common use case for doing so is to access the [Team Space](https://help.dropbox.com/teams-admins/team-member/team-space-overview), which are special team managed folders available to some Dropbox Business teams. Refer to the [DBX Team Files Guide](https://dropbox.com/lp/developers/reference/dbx-team-files-guide) for more information"

**Reply by m90** [reply]
Thanks for the digging. That's a bit mind bending, but explains what you're seeing I guess?

Since you already offered it, would you mind opening a PR that updates the docs to reflect these findings? Thank you.

**Reply by m90** [reply]
Looking at the code I can see `PathRoot` can be given to the Dropbox SDK:

```go
// Config contains parameters for configuring the SDK.
type Config struct {
	// OAuth2 access token
	Token string
	// Logging level for SDK generated logs
	LogLevel LogLevel
	// Logging target for verbose SDK logging
	Logger *log.Logger
	// Used with APIs that support operations as another user
	AsMemberID string
	// Used with APIs that support operations as an admin
	AsAdminID string
	// Path relative to which action should be taken
	PathRoot string
	// No need to set -- for testing only
	Domain string
	// No need to set -- for testing only
	Client *http.Client
	// No need to set -- for testing only
	HeaderGenerator func(hostType string, namespace string, route string) map[string]string
	// No need to set -- for testing only
	URLGenerator func(hostType string, namespace string, route string) string
}
```

I wonder if instead this tool would need to set a value for this to work consistently, but tbh I don't understand yet which value would be needed here.

**Reply by shalafi99** [reply]
> Thanks for the digging. That's a bit mind bending, but explains what you're seeing I guess?
> 
> Since you already offered it, would you mind opening a PR that updates the docs to reflect these findings? Thank you.

I suppose it does (explain it). About the PR, I do not mind it at all. Gotta confess it will be my first so it might take a little time while checking out the proper way to go about it.

**Reply by shalafi99** [reply]
> I wonder if instead this tool would need to set a value for this to work consistently, but tbh I don't understand yet which value would be needed here.

What do you think? I suppose the explanation could go both ways, somewhat like this (draft wording)

Other parameters
Important: If you chose App folder access during the creation of your Dropbox app in step 1 above, DROPBOX_REMOTE_PATH will be a relative path under the App folder (_For example, DROPBOX_REMOTE_PATH=/somedir means the backup file will be uploaded to /Apps/myapp/somedir_). On the other hand, if you chose Full Dropbox access the value for DROPBOX_REMOTE_PATH will be an absolute path inside your Dropbox storage (_Using the same value example above, the backup file would be uploaded to a folder "somedir" placed directly in the root for Dropbox_)

Too long?

**Reply by m90** [reply]
I think it's just about the right length. If you need help with getting that PR ready, let me know. Thanks a lot in advance in any case.

**Reply by shalafi99** [reply]
> I think it's just about the right length. If you need help with getting that PR ready, let me know. Thanks a lot in advance in any case.

Just one question, for starters: is it perfectly ok to not create an issue out of this discussion and instead just reference the PR directly to here?

**Reply by m90** [reply]
Yup, that's perfectly fine with me.

**Reply by shalafi99** [reply]
> Yup, that's perfectly fine with me.

I think I got it - PR #473
